content/blog/2023-12-11-A DSL for Metastability.md

```diff
@@ -0,0 +1,237 @@
+Note that I have heavily used sugaring, as you can imaging it is not pleasant to write actual code in the specified grammar.
+The sugaring should be self-explanatory.
+Let us first see the code, and then I will explain the scenario.
+
+```
+### Receive
+while not inq.is_full:
+    m = take();
+    put(inq, m);
+
+### Process
+if not is_processing {
+    t_t = t_t + 1;
+} else {
+    t_p = t_p + 1;
+}
+if t_p == T_p {
+    is_processing = false;
+    t_p = 0;
+    T_p = T_p0;
+} else {
+    nop;
+}
+if t_t == T_t {
+    is_processing = true;
+    t_t = 0
+    r = {type: "routing message", payload: "-", destination: BROADCAST};
+    if not outq.is_full {
+        put(outq, r);
+    } else {
+        nop;
+    }
+}
+while not inq.is_empty {
+    m = get(inq);
+    if m.type == "routing message" {
+        if is_processing {
+            T_p = T_p + 1;
+        } else {
+            nop;
+        }
+    } else if m.type == "update" {
+        is_processing = true;
+        t_t = 0
+        r = {type: "routing message", payload: "-", destination: BROADCAST};
+        if not outq.is_full {
+            put(outq, r);
+        } else {
+            nop;
+        }
+    } else {
+        nop;
+    }
+}
+
+### Send
+while not outq.is_empty:
+    m = get(outq);
+    give(m);
+```
+First, some notes:
+1. $\texttt{take}$ and $\texttt{give}$ are keywords to interact with the network. I did not write them in the grammar as I am still not sure how to treat them. Informally, $\texttt{take}$ receives the next request to be delivered by the network, and $\texttt{give}$ gives a request from the output queue to the network.
+2. $\texttt{BROADCAST}$ is a keywork to denote that this request must be sent to all of the agents.
+3. $r.f$ is sugar to access the field $f$ of request $r$, and $q.$is_full is sugar for checking the size of the queue (similarly for is_empty).
+
+The example here is from a [paper by Floyd and Van Jacobson](https://ee.lbl.gov/papers/sync_94.pdf).
+The paper studies synchronization of periodic routing messages.
+In a nutshell, this is what happens:
+1. Each agent, upon the expiration of its $T_t$ timer enters a listening state, during which it listens for routing messages from other nodes and processes them. Moreover, it sends them a routing message once it enters the listening state.
+2. For each message that it receives during the listening state, it spends an extra step in this state, captured by the increment to the timer for this state, $T_p$.
+3. Once the $T_p$ timer expires, the agent enters the idle state in which it is waiting for the expiration of the $T_t$ timer, captured by incrementing $t_t$. If it receives messages during this period, it just ignores them.
+4. There is a special message of the type "update", that can be injected anytime by the environment. This message abruptly takes the agent to the listening state, no matter the value of the timers.
```

Comment:
Here is a broader question that I have been wondering about all along, but which is especially salient given the example: where does the "initial input" for these programs come from? For example, it's clear how the program in the DSL reacts to messages and events produced by the environment. But where do the messages come from, originally?

I guess I think this is important because, of course, whether the program enters metastability depends on the way it is used. I understand that your long-term goal is to do some model-checking, i.e., to check whether there is _any_ condition under which the system enters metastability. But I think you're not there yet, and you're starting with just executing a _single_ run of the system to see if it exhibits the relevant conditions. So where does the initial stimulus for that single run come from?

Reply:
For now we have been crafting the initial conditions by hand, in order to capture some metastability examples. I think the initial conditions have to be defined via a specification of the system's desired behavior.

As for the inputs, they are usually packed into environment actions. For instance, the input load for a consumer service is captured by the environment just arbitrarily putting requests into the network, to be sent to some agent in the system.

Reply:
That makes sense! Could be worth briefly mentioning how that works in the post.

Reply:
Done.